codalab/worker_manager/slurm_batch_worker_manager.py

```diff
@@ -0,0 +1,268 @@
+        if self.args.dry_run:
+            return
+
+        batch_script = str(work_dir.joinpath(slurm_args['job-name'] + '.slurm'))
+        self.save_job_definition(batch_script, job_definition)
+        job_id_str = self.run_command([self.SBATCH_COMMAND, batch_script])
+
+        match = re.match(self.SBATCH_COMMAND_RETRUN_REGEX, job_id_str)
+        if match is not None:
+            job_id = match.group(1)
+        else:
+            logger.error("Cannot find job_id in {}.".format(job_id_str))
+            return
+
+        # Add the newly submitted job to submitted_jobs for tracking purpose
+        self.submitted_jobs.add(job_id)
+
+    def run_command(self, command):
+        proc = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output, errors = proc.communicate()
+        if output:
+            logger.info("Executed command: {}".format(' '.join(command)))
+            return output.decode()
+        if errors:
+            print(
+                "Failed to executing {}: {}: {}".format(' '.join(command), errors, proc.returncode)
+            )
+            logger.error(errors)
+        return ""
+
+    def save_job_definition(self, job_file, job_definition):
+        """
+        Save the batch job definition to file.
+        :param job_file: a file storing the Slurm batch job configuration
+        :param job_definition: the job definition of a Slurm batch job
+        :return:
+        """
+        with open(job_file, 'w') as f:
+            f.write(job_definition)
+        logger.info("Saved the Slurm Batch Job Definition to {}".format(job_file))
+
+    def create_job_definition(self, slurm_args, command):
+        """
+        Create a Slurm batch job definition structured as a list of Slurm batch arguments and a srun command
+        :param slurm_args: arguments for launching a Slurm batch job
+        :param command: arguments for starting a CodaLab worker
+        :return: a string containing the Slurm batch job definition
+        """
+        sbatch_args = [
+            '{} --{}={}'.format(self.SBATCH_PREFIX, key, slurm_args[key])
+            for key in sorted(slurm_args.keys())
+        ]
+        # Using the --unbuffered option with srun command will allow output
+        # appear in the output file as soon as it is produced.
+        srun_args = [self.SRUN_COMMAND, '--unbuffered'] + command
+        # job definition contains two sections: sbatch arguments and srun command
+        job_definition = '#!/bin/bash\n\n' + '\n'.join(sbatch_args) + '\n\n' + ' '.join(srun_args)
+        print("Slurm Batch Job Definition")
+        print(job_definition)
+        return job_definition
+
+    def create_random_job_name(self, job_name):
+        """
+        Generate a random Slurm job name
+        :param job_name:
+        :return: slurm job name
+        """
+        return self.username + "-" + job_name + '-' + str(random.randint(0, 5000000))
```

Comment:
since we're importing uuid anyway, maybe replace `str(random.randint(0, 5000000))` with a uuid?

Reply:
That's a good idea. Coz we don't actually those many random string. Do we want to cut the first few digits e.g. 8 or we want to use the full hex string (32 digits), which might be quite long?

Reply:
Yup, let's shorten.